staging/vhost-device-gpu/src/vhu_gpu.rs

```diff
@@ -551,30 +470,37 @@ impl VhostUserGpuBackend {
         virtio_gpu: &mut VirtioGpu,
         vring: &VringRwLock,
     ) -> Result<()> {
-        // Collect all pending requests
-        debug!("Processing control queue");
-        let Some(ref atomic_mem) = self.mem else {
-            return Err(Error::NoMemoryConfigured.into());
-        };
-        let requests: Vec<_> = vring
+        let mem = self.mem.as_mut().unwrap().memory().into_inner();
+        let desc_chains: Vec<_> = vring
             .get_mut()
             .get_queue_mut()
-            .iter(atomic_mem.memory())
-            .map_err(|_| Error::DescriptorNotFound)?
+            .iter(mem.clone())
+            .map_err(Error::CreateIteratorDescChain)?
             .collect();
 
-        let mem: GuestMemoryMmap = (*atomic_mem.memory().into_inner()).clone();
+        let mut signal_used_queue = false;
+        for desc_chain in desc_chains {
+            let head_index = desc_chain.head_index();
+            let mut reader = desc_chain
+                .clone()
+                .reader(&mem)
+                .map_err(Error::CreateReader)?;
+            let mut writer = desc_chain.writer(&mem).map_err(Error::CreateWriter)?;
+
+            self.process_control_queue_chain(
+                virtio_gpu,
+                vring,
+                head_index,
+                &mut reader,
+                &mut writer,
+                &mut signal_used_queue,
+            )?;
+        }
 
-        if self.process_requests(requests, virtio_gpu, vring).is_ok() {
-            // Send notification once all the requests are processed
-            debug!("Sending processed request notification");
-            vring
-                .signal_used_queue()
-                .map_err(|_| Error::NotificationFailed)?;
-            debug!("Notification sent");
+        if signal_used_queue {
+            vring.signal_used_queue().map_err(Error::SignalUsedQueue)?;
         }
 
-        debug!("Processing control queue finished");
```

Comment:
For now i think it'll be nice to leave these debug messages, and later when the Branch is ready we do a cleanup of unnecessary debug messages. WDYT

Reply:
Yes sure, or if we made some of the frequent messages `trace!` we could even leave them.